src/Common/src/Interop/Unix/System.Net.Security.Native/SecuritySafeHandles.cs

```diff
@@ -0,0 +1,98 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+using System.Text;
+using Microsoft.Win32.SafeHandles;
+
+namespace System.Net.Security
+{
+    internal sealed class SafeFreeNegoCredentials : SafeFreeCredentials
+    {
+        private SafeGssCredHandle _credential;
+
+        public SafeGssCredHandle GssCredential
+        {
+            get { return _credential; }
+        }
+
+        public SafeFreeNegoCredentials(string username, string password, string domain) : base(IntPtr.Zero, true)
+        {
+            bool ignore = false;
+            _credential = SafeGssCredHandle.Create(username, password, domain);
+            _credential.DangerousAddRef(ref ignore);
+        }
+
+        public override bool IsInvalid
+        {
+            get { return (null == _credential); }
+        }
+
+        protected override bool ReleaseHandle()
+        {
+            _credential.DangerousRelease();
+            _credential = null;
+            return true;
+        }
+    }
+
+    internal sealed class SafeDeleteNegoContext : SafeDeleteContext
+    {
+        private SafeGssNameHandle _targetName;
+        private SafeGssContextHandle _context;
+
+        public SafeGssNameHandle TargetName
+        {
+            get { return _targetName; }
+        }
+
+        public SafeGssContextHandle GssContext
+        {
+            get { return _context; }
+        }
+
+        public SafeDeleteNegoContext(SafeFreeNegoCredentials credential, string targetName)
+            : base(credential)
+        {
+            try
+            {
+                _targetName = SafeGssNameHandle.CreatePrincipal(targetName);
+            }
+            catch
+            {
+                Debug.Assert((null != credential), "Null credential in SafeDeleteNegoContext");
+                credential.DangerousRelease();
```

Comment:
This is problematic.  Consider what happens here.  We call the base ctor, which calls AddRef.  We then run the derived ctor, which gets an exception and calls DangerousRelease... but the base still initialized, and its IsInvalid will return false because _credential was set.  Now there are two issues.  The minor one is that we've created a SafeHandle that's going to get finalized, which has a cost.  The much worse one is that when it does get finalized, it's going to call ReleaseHandle, which is going to call DangerousRelease again!

Given how you've got all of this arranged, seems like the best option with the current structure would be to simply call Dispose here rather than calling credential.DangerousRelease.  That should properly let both this and the base do the "right thing"... it fixes both mentioned problems, as well as the oddity of calling DangerousRelease here when there's no visible corresponding AddRef.

In addition, I do think we need to reevaluate whether there's the potential to change what the PAL is such that when there isn't actually a native resource to be wrapped that we don't need to artifically create a SafeHandle hierarchy for one.